### PR TITLE
`driver`: filter orders based on user balances

### PR DESCRIPTION
**Pull request by cowfee**
Progress on https://github.com/cowprotocol/services/issues/1672.

Filters orders based on trader balances. If the traders don't have enough balance to cover the order, it will be removed before being sent to the solver.

TODO after merging:
- Configure ethflow address in infra.
- Extend the test.

### Test Plan

Added automated test.

### CODE REVIEW

File: crates/driver/src/domain/competition/auction.rs [hunk 1]
```diff
@@ -76,12 +62,95 @@ impl Auction {
...
+            .flat_map(|order| {
+                [
+                    (order.trader(), order.sell.token),
+                    (order.trader(), order.buy.token),
+                ]
+            })
+            .unique()
+            .collect_vec();
+        let mut balances: HashMap<(order::Trader, eth::TokenAddress), eth::TokenAmount> =
+            join_all(tokens_by_trader.into_iter().map(|(trader, token)| {
```
Comment (vkgnosis) [inline crates/driver/src/domain/competition/auction.rs:108]: The same trader token pair will be fetched multiple times here. Maybe collect tokens_by_trader into a set instead of a vec.

File: crates/driver/src/domain/competition/auction.rs [hunk 2]
```diff
@@ -76,12 +62,95 @@ impl Auction {
...
+            match sell_balance.0.checked_sub(order.sell.amount.into()) {
+                Some(remaining) => sell_balance.0 = remaining,
+                None => return false,
+            }
+
+            // Buy amounts are deposited into the trader's balance when the order settles.
+            let buy_balance = balances
+                .get_mut(&(order.trader(), order.buy.token))
+                .unwrap();
+            buy_balance.0 = buy_balance.0.saturating_add(order.buy.amount.into());
```
Comment (vkgnosis) [inline crates/driver/src/domain/competition/auction.rs:146]: This doesn't work because all transfers happen at the start of the settlement before any additional buy amount shows up.

File: crates/driver/src/domain/competition/auction.rs [hunk 3]
```diff
@@ -76,12 +62,95 @@ impl Auction {
...
+            .into_iter()
+            .collect();
+
+        // Filter out orders which the trader doesn't have enough balance to pay for.
+        self.orders.retain(|order| {
+            // Partial orders don't require the trader to hold the full balance, since they
+            // don't have to be fully fulfilled.
+            if order.is_partial() {
+                return true;
+            }
```
Comment (vkgnosis) [inline crates/driver/src/domain/competition/auction.rs:129]: The problem then is that the following can happen:
1.Solver settles partially fillable order for some sell amount.
2. Solver settles FOK order for some some sell amount.
3. User doesn't have enough balances for both fills and settlement fails.

**Review by vkgnosis — COMMENTED** [review]
This code does some things differently than the code in `crates/solver/src/order_balance_filter.rs` but the goal of both modules is the same so the differences should be explained.

The goal of the module (IMO) is to ensure that no matter what orders a solver picks, the resulting settlement will not be invalid because a picked order doesn't have enough balance. In order to do that we need to be pessimistic about balance assignment.

**Comment by vkgnosis** [comment]
This is ready for review now. The test needs to be improved to test more of the cases we care about. I didn't do this because I am uncertain about how to organize the test code right and I don't want that to block the main functionality.

File: crates/driver/src/domain/competition/auction.rs [hunk 4]
```diff
@@ -53,33 +90,114 @@ impl Auction {
             };
             std::cmp::Reverse((
                 class,
+                // TODO No need to expose Tokens anymore, just use the auction
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/auction.rs:93]: This TODO was taken care of, right?

File: crates/driver/src/domain/competition/mod.rs [hunk 5]
```diff
@@ -46,7 +46,9 @@ pub struct Competition {
 
 impl Competition {
     /// Solve an auction as part of this competition.
-    pub async fn solve(&self, auction: &Auction) -> Result<Solved, Error> {
+    pub async fn solve(&self, auction: Auction) -> Result<Solved, Error> {
+        let auction = auction.prioritize(&self.eth).await;
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/mod.rs:50]: Looks like changing the auction after passing it into this function results in a few edge cases.
Could this be done before hand instead?

File: crates/driver/src/tests/cases/order_prioritization.rs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        .solution(ab_solution())
+        .done()
+        .await;
+
+    // Only check that the solve endpoint can be called successfully, which means
+    // that the solver received the orders sorted.
+    test.solve().await.ok();
+}
+
+// TODO Comment
```
Comment (MartinquaXD) [inline crates/driver/src/tests/cases/order_prioritization.rs:35]: Left over TODO

File: crates/driver/src/domain/competition/auction.rs [hunk 7]
```diff
@@ -53,33 +90,114 @@ impl Auction {
...
+            ((trader, token), balance)
+        }))
+        .await
+        .into_iter()
+        .collect();
+
+        self.orders.retain(|order| {
+            // TODO: We should use balance fetching that takes interactions into account
+            // from `crates/shared/src/account_balances/simulation.rs` instead of hardcoding
+            // an Ethflow exception. https://github.com/cowprotocol/services/issues/1595
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/auction.rs:120]: I believe we already have the pre-interaction aware balance fetcher, right?
In that case it might be less effort to immediately use that instead of first introducing all the ethflow machinery for this exception and later dropping that again.

**Comment by nlordell** [comment]
This still needs to be merged for co-location.

File: crates/driver/src/domain/competition/auction.rs (hunk 7 above)
Comment (MartinquaXD) [reply crates/driver/src/domain/competition/auction.rs:119]: Leaving that unaddressed for now as things are not as straight forward than I thought.

File: crates/driver/src/domain/competition/auction.rs [hunk 8]
```diff
@@ -23,26 +34,52 @@ pub struct Auction {
...
         let tokens = Tokens(tokens.map(|token| (token.address, token)).collect());
 
         // Ensure that tokens are included for each order.
+        let weth = eth.contracts().weth_address();
         if !orders.iter().all(|order| {
             tokens.0.contains_key(&order.buy.token.wrap(weth))
                 && tokens.0.contains_key(&order.sell.token.wrap(weth))
         }) {
-            return Err(InvalidTokens);
+            return Err(Error::InvalidTokens);
```
Comment (fleupold) [inline crates/driver/src/domain/competition/auction.rs:52]: nit: may be good to at least log which tokens are missing (I imagine this being otherwise very complex to debug)

File: crates/driver/src/domain/competition/auction.rs [hunk 9]
```diff
@@ -23,26 +34,52 @@ pub struct Auction {
...
+            orders,
+            tokens,
+            gas_price: eth.gas_price().await?,
+            deadline,
+        })
+    }
+
+    /// [`None`] if this auction applies to a quote. See
+    /// [`crate::domain::quote`].
+    pub fn id(&self) -> Option<Id> {
```
Comment (fleupold) [inline crates/driver/src/domain/competition/auction.rs:66]: Separate, but solvers have been asking us to also send IDs for quote instances (so they can easily debug and get back to us with identifiable instances). One idea we had was to "pre-generate" the quote_id and then attach it to each request. This will even allow solvers to check independently what they quoted for a given order (assuming we expose quote_id in our API)

File: crates/driver/src/domain/competition/auction.rs [hunk 10]
```diff
@@ -55,31 +92,111 @@ impl Auction {
...
+                        order::Partial::No => false,
+                    },
+                })
+                .context("Remaining::from_order")?;
+                let sell = remaining
+                    .remaining(order.sell.amount.0)
+                    .context("remaining_sell")?;
+                let fee = remaining
+                    .remaining(order.fee.user.0)
+                    .context("remaining_fee")?;
```
Comment (fleupold) [inline crates/driver/src/domain/competition/auction.rs:165]: Honestly it's not really clear to me what `remaining.remaining` is supposed to do, but I guess one is expected to dig into the code to understand this?

File: crates/driver/src/domain/competition/mod.rs (hunk 5 above)
Comment (fleupold) [reply crates/driver/src/domain/competition/mod.rs:50]: Any more thoughts on this?

File: crates/driver/src/infra/config/file/mod.rs [hunk 11]
```diff
@@ -178,6 +178,11 @@ pub struct ContractsConfig {
 
     /// Override the default address of the WETH contract.
     pub weth: Option<eth::H160>,
+
+    /// Sets the Ethflow contract address. Without this we cannot detect Ethflow
+    /// orders, which leads to such orders not being solved because it appears
+    /// that the user doesn't have enough sell token balance.
+    pub ethflow: Option<eth::H160>,
```
Comment (fleupold) [inline crates/driver/src/infra/config/file/mod.rs:185]: Should this be mandatory?

**Review by fleupold — APPROVED** [review]
Logic looks good to me, although a lot of the architectural design here is still quite new to me.

File: crates/driver/src/domain/competition/auction.rs (hunk 10 above)
Comment (MartinquaXD) [reply crates/driver/src/domain/competition/auction.rs:165]: Ideally you shouldn't have to dig into the code. But the naming for this type and function is not straight forward.
Basically the idea is that the `remaining` variable holds all the information needed to know how much of an order was executed. And the `remaining` method then allows you to scale the initial amount to the amount that is currently left in the order.

File: crates/driver/src/infra/config/file/mod.rs (hunk 11 above)
Comment (MartinquaXD) [reply crates/driver/src/infra/config/file/mod.rs:185]: Yeah, we can make this mandatory. We just have to make sure the required infra changes are already there before merging the PR to avoid panics.

File: crates/driver/src/domain/competition/auction.rs (hunk 8 above)
Comment (nlordell) [reply crates/driver/src/domain/competition/auction.rs:52]: Personally, I'm OK with no log in the long run:
- We get a `InvalidTokens` error
- We look at the stored auction JSON from the autopilot.

Now, that being said - I don't think we are saving the auction JSON that is being sent to the drivers for every run. I really think we should in the context of co-location (as it would want to store everything we send to externally operated `driver`s). Once we do store these auction JSON documents, however, I think the extra log here isn't super useful (you will always check the auction JSON anyway to check if the missing token is indeed missing, so I don't think it saves you much time).

File: crates/driver/src/domain/competition/auction.rs (hunk 9 above)
Comment (nlordell) [reply crates/driver/src/domain/competition/auction.rs:66]: I _think_ some solvers use the `id` field to disambiguate between a `solve` and a `quote` (I think Quasimodo does this specifically). This might be a backwards incompatible change (just warning).

I 100% agree we should do this, however.

Also note that this `id` is the id of an auction to `solve`. Driver's have a separate API for handling quotes.

File: crates/driver/src/domain/competition/auction.rs [hunk 12]
```diff
@@ -23,26 +34,52 @@ pub struct Auction {
...
+    /// The orders for the auction.
+    pub fn orders(&self) -> &[competition::Order] {
+        &self.orders
+    }
+
+    // Prioritize the orders such that those which are more likely to be fulfilled
+    // come before less likely orders. Filter out orders which the trader doesn't
+    // have enough balance to pay for.
+    //
+    // Prioritization is skipped during quoting. It's only used during competition.
```
Comment (nlordell) [inline crates/driver/src/domain/competition/auction.rs:79]:
```suggestion
    /// Prioritize the orders such that those which are more likely to be fulfilled
    /// come before less likely orders. Filter out orders which the trader doesn't
    /// have enough balance to pay for.
    ///
    /// Prioritization is skipped during quoting. It's only used during competition.
```

File: crates/driver/src/domain/competition/auction.rs [hunk 13]
```diff
@@ -23,26 +34,52 @@ pub struct Auction {
...
+    pub async fn prioritize(mut self, eth: &Ethereum) -> Self {
```
Comment (nlordell) [inline crates/driver/src/domain/competition/auction.rs:80]: This comment makes me think that we should have 2 auction types:
1. that represents the `Auction` we get for the protocol to solve
2. the `Augmented` auction that is the result of:
  * fetching token data
  * fetching liquidity
  * prioritization and filtering
  
If we do have to separate types, then it becomes clear at a type level which "methods" are applied on an `Auction` for solving and for quoting:
* `async Auction -> auction::Augmented` transformation does order prioritization.
* `async quote::Order -> auction::Augmented` transformation does **not** do order prioritization.

**Not a suggestion for this PR**, though. Just a general suggestion to clean up (what is IMO) a code smell that is starting to appear as we add some more logic to these domain types.

File: crates/driver/src/domain/competition/auction.rs [hunk 14]
```diff
@@ -55,31 +92,111 @@ impl Auction {
...
+        // Has to be separate closure due to compiler bug.
+        let f = |order: &competition::Order| -> (order::Trader, eth::TokenAddress) {
+            (order.trader(), order.sell.token)
+        };
+        let tokens_by_trader = self.orders.iter().map(f).unique();
+        let mut balances: HashMap<
+            (order::Trader, eth::TokenAddress),
+            Result<eth::TokenAmount, crate::infra::blockchain::Error>,
+        > = join_all(tokens_by_trader.map(|(trader, token)| async move {
+            let balance = eth.balance_of(trader.into(), token).await;
```
Comment (nlordell) [inline crates/driver/src/domain/competition/auction.rs:109]: We would need to pull in the `BalanceFetching` implementation where we simulate with hooks here to not break the current hooks implementation.

File: crates/driver/src/infra/blockchain/contracts.rs [hunk 15]
```diff
@@ -46,6 +57,10 @@ impl Contracts {
     pub fn weth_address(&self) -> eth::WethAddress {
         self.weth.address().into()
     }
+
+    pub fn ethflow_address(&self) -> Option<eth::ContractAddress> {
+        self.ethflow
+    }
```
Comment (nlordell) [inline crates/driver/src/infra/blockchain/contracts.rs:63]: This wouldn't be needed with hooks support right? Ideally, we would keep this out of the `driver`.

File: crates/driver/src/infra/config/file/mod.rs (hunk 11 above)
Comment (nlordell) [reply crates/driver/src/infra/config/file/mod.rs:185]: I would like to avoid this configuration change if possible. I think it should be possible if we correctly simulate balances with pre-interactions for orders.

**Review by nlordell — APPROVED** [review]
LGTM.

Two comments that I think should lead to follow ups:
- I think we need two `Auction` types - one for the auction we get from the protocol, and one for the augmented auction to send to solvers.
- We need to use the `BalanceFetching` implementation that supports simulations of token balances with pre/post interactions. This would allow us to drop `ethflow` specific conditions and logic (which I would like to keep out of the driver).

File: crates/driver/src/infra/blockchain/contracts.rs (hunk 15 above)
Comment (MartinquaXD) [reply crates/driver/src/infra/blockchain/contracts.rs:63]: Although the ethflow contract holds enough `ETH` for all the trades it may not hold enough `WETH` for all the trades. So unless I'm missing something this is needed unfortunately.
I guess one thing to get around this is to use the balance fetching implementation that considers pre-interactions for all orders that need them and not updating the user's regular token balances when including orders because we don't know whether the pre-interaction would actually use some of the user's balance.

Comment (nlordell) [reply crates/driver/src/infra/blockchain/contracts.rs:63]: Hmm, I'm not sure I understand. Ethflow orders use pre-interactions that wrap `WETH`. So if we simulate the `WETH` balance after executing the order's pre-interaction, we will get sufficient `WETH` to trade all user's Ethflow orders (since it always wraps all of its ETH).

Comment (MartinquaXD) [reply crates/driver/src/infra/blockchain/contracts.rs:63]: The point I was trying to make was more general. Right now we: 
* collect all tuples `(owner, sell_token)`
* fetch these balances
* go through all orders and decrement the corresponding user balance until nothing is left

If we want to use the balance fetcher we have to simulate the balance of each order individually because grouping them doesn't really make sense because they could have vastly different pre-interactions which get the required `sell_token` balance.
That's why I suggested to keep doing what are currently doing for orders that don't use any pre-interactions and for all the other orders we could simply check if the user has the expected balance after simulating the pre-interaction.
But this could be problematic when a user has 10 WETH, 1 order selling 10 WETH and 1 order selling 10 WETH with some random pre-interaction.

The point I'm trying to make is that we have way less guarantees about sell token balances with orders that include pre-interactions.

Comment (MartinquaXD) [reply crates/driver/src/infra/blockchain/contracts.rs]: BTW I tried going down the road of making pre-interaction aware balance fetching available in the `driver` but I think I first have to implement the old `CodeSimulating` trait for the driver `infra::Simulator` and for that I think it would be beneficial to move `ethrpc` into its own crate first.

**Comment by MartinquaXD** [comment]
Merging to unblock further roll out for now.
Will follow up with the refactoring required to drop the ethflow edge cases.
